Add basic how-to for Java toolchain using the same structure as the Rust guide (#7)
canonical/ubuntu-for-developers-docs: docs/howto/java-use.md

If you want to use `javac` directly, refer to the example [Using `javac` directly](#how-to-use-javac-directly).

## Install Java Development Kit

Run

```
apt install default-jdk
```

to install the default Java Development Kit for your Ubuntu release.

Use ``apt install openjdk-21-jdk`` to install the latest LTS release.

## Install Maven

Run

```
apt install maven
```

to install Maven and the default Java Development Kit from the Ubuntu archive.

Alternatively, download Maven from https://maven.apache.org/download.cgi.

## Create a Java project using Maven

1. Create a new Java project using the `archetype:generate` Maven sub-command:

    ```
    mvn archetype:generate -DgroupId=com.yourcompany -DartifactId=helloworld -Dversion=1.0-SNAPSHOT -Dpackage=com.yourcompany.helloworld -DarchetypeGroupId=org.apache.maven.archetypes -DarchetypeArtifactId=maven-archetype-quickstart -DarchetypeVersion=1.4
    ```

>     Press `Enter` when prompted to confirm your selection.

Reviewer: ```suggestion
    Press <kbd>Enter</kbd> when prompted to confirm your selection.
```

    This creates a new project using [maven-archetype-quickstart](https://maven.apache.org/archetypes/maven-archetype-quickstart/).

2. Change to the project directory:

    ```
    cd hello_world
    ```

    Maven has already set up a basic project structure that includes a `Hello world` application and a unit test:

    ```
    helloworld
    |-- pom.xml
    `-- src
        |-- main
        |   `-- java
        |       `-- $package
        |           `-- App.java
        `-- test
            `-- java
                `-- $package
                    `-- AppTest.java
    ```

>     ```
>     cat src/main/java/com/yourcompany/helloworld/App.java
>     ```
> 
>     ```Java

Reviewer: ```suggestion
    ```{code-block} java
    :caption: src/main/java/com/yourcompany/helloworld/App.java

```

    package com.yourcompany.helloworld;

    /**
    * Hello world!
    *
    */
    public class App
    {
        public static void main( String[] args )
        {
            System.out.println( "Hello World!" );
        }
    }
    ```

>     ```
>     cat src/test/java/com/yourcompany/helloworld/AppTest.java
>     ```
> 
>     ```Java

Reviewer: ```suggestion
    ```{code-block} java
    :caption: src/test/java/com/yourcompany/helloworld/AppTest.java

```

    package com.yourcompany.helloworld;

    import static org.junit.Assert.assertTrue;

    import org.junit.Test;

    /**
    * Unit test for simple App.
    */
    public class AppTest
    {
        /**
        * Rigorous Test :-)
        */
        @Test
        public void shouldAnswerWithTrue()
        {
            assertTrue( true );
        }
    }
    ```

3. Build and run the program:

    ```
    mvn -Dmaven.compiler.release=8 package
    ```

>     Notice: -Dmaven.compiler.release=8. Maven quickstart generates a project that targets Java 7 which is no longer supported by the Java 21 LTS release. The project target can be changed by updating the `maven.compiler.target` and `maven.compiler.source` properties in `pom.xml`

Reviewer: ```suggestion
    :::{note}
    Notice the `-Dmaven.compiler.release=8` option.
    Maven quickstart generates a project that targets Java 7 which is no longer supported by the Java 21 LTS release. The project target can be changed by updating the `maven.compiler.target` and `maven.compiler.source` properties in `pom.xml`
    :::
```

    This builds and runs unit tests.

    Run the application:

    ```
    java -cp target/helloworld-1.0-SNAPSHOT.jar com.yourcompany.helloworld.App
    Hello World!
    ```

## Install Gradle

Download Gradle from [gradle.org](https://gradle.org/releases).

> Note: Gradle introduced Java 21 support in version 8.5.

Reviewer: ```suggestion
:::{note}
Gradle introduced Java 21 support in version 8.5.
:::
```

Alternatively, run

```
snap install gradle
```

to install community-maintained Gradle snap. The snap provides Gradle version 7 but does not support Java 21.

> Notice: The snap requires setting up the `JAVA_HOME` variable. For example:
> 
> ```
> export JAVA_HOME=/usr/lib/jvm/java-21-openjdk-amd64
> ```

Reviewer: ````suggestion
:::{note}
The snap requires setting up the `JAVA_HOME` variable. For example:

```bash
export JAVA_HOME=/usr/lib/jvm/java-21-openjdk-amd64
```
:::
````

## Create a Java project using Gradle

1. Create a Java project using the `gradle init` command:

>     ```

Reviewer: ```suggestion
    ```bash
```

    mkdir helloworld
    cd helloworld
    gradle init \
    --type java-application \
    --dsl kotlin \
    --test-framework junit-jupiter \
    --package com.yourcompany.helloworld \
    --project-name helloworld  \
    --no-split-project \
    --no-incubating
    ```

>     Press `Enter` when prompted for the Java version.

Reviewer: ```suggestion
    Press <kbd>Enter</kbd> when prompted for the Java version.
```

    Gradle sets up a basic project structure with a `Hello World!` application:

>     ```
>     cat app/src/main/java/com/yourcompany/helloworld/App.java
>     ```
> 
>     ```Java

Reviewer: ```suggestion
    ```{code-block} java
    :caption: app/src/main/java/com/yourcompany/helloworld/App.java

```

    /*
    * This Java source file was generated by the Gradle 'init' task.
    */
    package com.yourcompany.helloworld;

    public class App {
        public String getGreeting() {
            return "Hello World!";
        }

        public static void main(String[] args) {
            System.out.println(new App().getGreeting());
        }
    }
    ```

>     ```
>     cat app/src/main/java/com/yourcompany/helloworld/App.java
>     ```
> 
>     and a unit test:
> 
>     ```
>     cat app/src/test/java/com/yourcompany/helloworld/AppTest.java
>     ```
> 
>     ```Java

Reviewer: ```suggestion

    and a unit test:

    ```{code-block} java
    :caption: app/src/test/java/com/yourcompany/helloworld/AppTest.java
```

    /*
    * This Java source file was generated by the Gradle 'init' task.
    */
    package com.yourcompany.helloworld;

    import org.junit.jupiter.api.Test;
    import static org.junit.jupiter.api.Assertions.*;

    class AppTest {
        @Test void appHasAGreeting() {
            App classUnderTest = new App();
            assertNotNull(classUnderTest.getGreeting(), "app should have a greeting");
        }
    }
    ```

    2. Build and run the project:

    ```
    ./gradlew run
    ```

    This builds and runs the project.

> ## How to Use `javac` Directly

Reviewer: ```suggestion
(how-to-use-javac-directly)=
## How to Use `javac` Directly
```

I think there might be a bug somewhere in Sphinx, because your implicit links should work, but Sphinx errors out when you use them.

1. Create a Hello World application in a file `App.java`:

    ```Java
    public class App {
        public String getGreeting() {
            return "Hello World!";
        }

        public static void main(String[] args) {
            System.out.println(new App().getGreeting());
        }
    }
    ```

    Run

    ```
    javac App.java -d out
    ```
    to compile the class file in the `out` directory.

    Execute the program:

    ```
    java -cp out/ App
    Hello World!
    ```

## Running Java application as shebang script

1. Create a Hello World application in a file `App`:

>     ```
>     #!/usr/bin/java
>     public class App {
>         public String getGreeting() {
>             return "Hello World!";
>         }
> 
>         public static void main(String[] args) {
>             System.out.println(new App().getGreeting());
>         }
>     }
>     ```
>     Note: the file lacks a .java extension.

Reviewer: ```suggestion
    ```{code-block}
    :caption: App

    #!/usr/bin/java
    public class App {
        public String getGreeting() {
            return "Hello World!";
        }

        public static void main(String[] args) {
            System.out.println(new App().getGreeting());
        }
    }
    ```

:::{important}
This file does not have a `.java` extension!
:::
```

Not annotating the code snippet as Java, the Sphinx syntax parser is yet to support the shebang injection in Java.

    Make the file executable:

>     ```

Reviewer: ````suggestion
    ```none
````

    chmod +x App
    ```

    Run the application

>     ```

Reviewer: ````suggestion
    ```none
````

            assertNotNull(classUnderTest.getGreeting(), "app should have a greeting");
        }
    }
    ```

    2. Build and run the project:

    ```
    ./gradlew run
    ```

    This builds and runs the project.

(how-to-use-javac-directly)=
## How to Use `javac` Directly

1. Create a Hello World application in a file `App.java`:

    ```Java
    public class App {
        public String getGreeting() {
            return "Hello World!";
        }

        public static void main(String[] args) {
            System.out.println(new App().getGreeting());
        }
    }
    ```

    Run

    ```
    javac App.java -d out
    ```
    to compile the class file in the `out` directory.

    Execute the program:

    ```
    java -cp out/ App
    Hello World!
    ```

## Running Java application as shebang script

1. Create a Hello World application in a file `App`:

    ```{code-block}
    :caption: App

    #!/usr/bin/java
    public class App {
        public String getGreeting() {
            return "Hello World!";
        }

        public static void main(String[] args) {
            System.out.println(new App().getGreeting());
        }
    }
    ```

> :::{important}
> This file does not have a `.java` extension!
> :::

Reviewer: ```suggestion
    :::{important}
    This file does not have a `.java` extension!
    :::
```

Needs to be indented to avoid breaking the MyST parser.
